packages/itwin/tree-widget/README.md

```diff
@@ -169,69 +209,66 @@ In general, the component is expected to be used by advanced users to inspect co
 
-Where `{tree}` specifies which tree component the feature is of.
+### Hierarchy level filtering
 
-Example:
+All tree components in this package allow users to filter nodes at each hierarchy level. The filter is applied to a single hierarchy level, which allows users to reduce amount of nodes being loaded - this is especially useful when a [hierarchy level size limit](#hierarchy-level-size-limiting) is hit:
 
-```ts
-import { UiItemsManager } from "@itwin/appui-react";
-import { TreeWidgetUiItemsProvider, ModelsTreeComponent } from "@itwin/tree-widget-react";
-...
-UiItemsManager.register(
-  new TreeWidgetUiItemsProvider({
-    defaultPanelLocation: StagePanelLocation.Left,
-    defaultPanelSection: StagePanelSection.End,
-    defaultTreeWidgetPriority: 1000,
-    onPerformanceMeasured={(feature, elapsedTime) => {
-      telemetryClient.log(`TreeWidget [${feature}] took ${elapsedTime} ms`);
-    }},
-    trees: [{
-      id: ModelsTreeComponent.id,
-      getLabel: ModelsTreeComponent.getLabel,
-      render: (props) => <ModelsTreeComponent { ...props } />,
-    }];
-  })
-);
-```
+![Hierarchy level filtering example](./media/hierarchy-level-filtering-example.gif)
 
-For individual tree components the `onPerformanceMeasured` callback should be supplied through props:
+## Telemetry
 
-```ts
-return (
-  <ModelsTreeComponent
-    onPerformanceMeasured={(feature, elapsedTime) => {
-      console.log(`TreeWidget [${feature}] took ${elapsedTime} ms`)
-    }}
-  />
-);
-```
+### Performance tracking
 
-## Usage tracking
+Components from this package allows consumers to track performance of specific features.
+
+This can be achieved by passing `onPerformanceMeasured` function to `CategoriesTreeComponent`, `ModelsTreeComponent`, `IModelContentTreeComponent` or `TreeWidgetUiItemsProvider`. The function is invoked with feature id and time elapsed as the component is being used. List of tracked features:
+
+- `"{tree}-initial-load"` - time it takes to load initial nodes after the tree is created.
+- `"{tree}-hierarchy-level-load"` - time it takes to load child nodes when a node is expanded.
+- `"{tree}-reload"` - time it takes to reload the tree after data in the iModel changes or it's being reloaded due to filtering.
+
+Where `{tree}` specifies which tree component the feature is of.
+
+### Usage tracking
 
 Components from this package allows consumers to track the usage of specific features.
 
 This can be achieved by passing `onFeatureUsed` function to `CategoriesTreeComponent`, `ModelsTreeComponent`, `IModelContentTreeComponent` or `TreeWidgetUiItemsProvider`. The function is invoked with feature id as the component is being used. List of tracked features:
 
 - `"choose-{tree}"` - when a tree is selected in the tree selector.
-- `"use-{tree}"` - when an interaction with a tree hierarchy happens.
+- `"use-{tree}"` - when an interaction with a tree hierarchy happens (a node is expanded or collapsed).
```

Comment:
This is now not entirely correct as this feature is also reported when other interactions happen (selecting nodes, changing visibility, filtering, etc.)

Reply:
Hm, I must have missed the change. Was it on purpose?

Reply:
No, it was that way from the start. I asked you whether it should be reported on header button actions and we decided that it is better to only report interactions with the hierarchy, which I assumed includes all interactions.

Reply:
This definitely shouldn't include interactions with header buttons. Since we already have other feature IDs for visibility and filtering, I assumed this one is used only for expand / collapse. But I guess it's okay to include all interactions, as long as we're consistent and don't change the meaning of what we report.

I'll update the readme.